gix-blame/src/file/function.rs

```diff
@@ -211,11 +211,14 @@ fn pass_blame_from_to(from: ObjectId, to: ObjectId, hunks_to_blame: &mut Vec<Unb
 /// Convert each of the unblamed hunk in `hunks_to_blame` into a [`BlameEntry`], consuming them in the process.
 /// `suspect` is expected to be present in the suspect-map in each [`UnblamedHunk`].
```

Comment:
With this change it appears the comment needs to be updated to reflect the fact that we don’t expect `suspect` to be in `suspects` any longer.

Reply:
Thanks, good catch - it will be fixed next time I push.

Please take a look at the updated comment above, it's my journey of trying to make it faster. But Git still has pixie dust that I simply can't replicate. Somehow it avoids a lot of work probably while defying physics.

Reply:
Well, sounds like it’s time to turn to my annotated version of `git blame` again. Maybe, at some point, it will tell its secrets. :-)

Reply:
It would be really interesting if there were statistics about the object database. How many objects did it actually read? How long were typical pack delta chains?

The latter information is actually available in `gitoxide`, I just never wired it up 

Either way, experiments with the pack-delta cache showed that the cache efficiency isn't getting better after a little more than 100MB, showing that everything that could be cached and reused was cached. One can go up to 2.1g pack delta cache, and one will still get the same amount of hits as before. So for this workload, one will have to read 116558 pack entries, and there is no way around it apparently.

`update_tree_entry` in Git takes 115ms, it's parsing tree entries. The respective version in `gitoxide` runs for 554ms. The key here seems to be `skip_uninteresting`, which consumes most of the time. I think a key is that it knows about the sort-order, so when it's past a particular entry it knows that a match cannot be found anymore.
 `gitoxide` always has to search all entries in the tree, so spends a lot more time decoding *everything* in a tree.

The physics-defying thing still is that apparently it doesn't spend much time in the object database at all, something that costs `gitoxide` tremendous amounts of time.

<img width="1207" alt="Screenshot 2024-12-25 at 20 23 08" src="https://github.com/user-attachments/assets/02a0a516-edde-4d31-982f-b33e9a9eda80" />

Reply:
From what I can tell, `git` seems to lazily decode tree entries (and thus maybe trees?). `skip_uninteresting` sequentially calls `update_tree_entry` which calls `decode_tree_entry` under the hood until a blob that matches the given path (and file mode as far as I can tell) is found. `skip_uninteresting` seems to go through the available entries roughly in alphabetical order. This is what I observed running a `printf`-instrumented version of `git blame STABILITY.md`. Now I’m going to have a look at a larger repository. :-)

Reply:
And `decode_tree_entry` appears to do almost nothing unless I’m missing something: https://github.com/git/git/blob/6258f68c3c1092c901337895c864073dcdea9213/tree-walk.c#L16-L47

Reply:
I also have [some additional results](https://github.com/GitoxideLabs/gitoxide/pull/1453#issuecomment-2561986611), which I think explain the pixie-dust with science🧬 😁.

It's notable that this is not necessarily the total amount of objects that are fetched, but it's the objects that have to be read once from a pack.